toolRepository-remote/src/main/scala/com/codacy/toolRepository/remote/ToolRepositoryRemote.scala

```diff
@@ -0,0 +1,145 @@
+package com.codacy.toolRepository.remote
+
+import akka.stream.Materializer
+import akka.stream.scaladsl.Sink
+import com.codacy.analysis.clientapi.definitions
+import com.codacy.analysis.clientapi.definitions.{PaginationInfo, PatternListResponse, ToolListResponse}
+import com.codacy.analysis.clientapi.tools.{ListPatternsResponse, ListToolsResponse, ToolsClient}
+import com.codacy.analysis.core.model.{AnalyserError, ParameterSpec, PatternSpec, ToolSpec}
+import com.codacy.analysis.core.tools.ToolRepository
+import com.codacy.plugins.api.languages.{Language, Languages}
+import com.codacy.toolRepository.remote.storage.{PatternSpecDataStorage, ToolSpecDataStorage}
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+class ToolRepositoryRemote(toolsClient: ToolsClient,
+                           toolsStorage: String => ToolSpecDataStorage,
+                           patternStorage: String => PatternSpecDataStorage)(implicit val ec: ExecutionContext,
+                                                                             implicit val mat: Materializer)
+    extends ToolRepository {
+
+  private val toolStorageFilename = "tools"
+  private val toolStorageInstance = toolsStorage(toolStorageFilename)
+
+  override lazy val list: Either[AnalyserError, Seq[ToolSpec]] = {
```

Comment:
I was expecting that the "Repository" would have some kind of "lazy val" on the list / get methods, similar to what we do on codacy-tools, so we don't need to fetch it multiple times.

As it is, we will be fetching every time we use it, but we create the "ToolRepositoryRemote" only once, which we should make sure that fetches things from the API only once, correct? Or am I missing something?

Reply:
Not sure I understood your point 🤔 . That list is a lazy val. And the get is a search over that lazy val. Were you thinking of a different approach?

Reply:
wow, maybe I am drunk, I could swear it was a def, don't mind me kind sir 😅